File: java/leetcode_grind/Day341.java
package leetcode_grind;

import java.util.Stack;

public class Day341 {
    // https://leetcode.com/problems/basic-calculator-iii/description/
    static class Solution1 {
        public int calculate(String s) {
            var eval = new Object() {
                int apply(char op, int x, int y) {
                    if (op == '+') {
                        return x;
                    } else if (op == '-') {
                        return -x;
                    } else if (op == '*') {
                        return x * y;
                    }
                    return x / y;
                }
            };

            s += "@";
            var stack = new Stack<Object>();
            var operator = '+';
            var current = 0;

            for (var ch : s.toCharArray()) {
                if (ch == ' ') {
                    continue;
                }

                if (Character.isDigit(ch)) {
                    current = current * 10 + (ch - '0');
                } else if (ch == '(') {
                    stack.push(operator);
                    operator = '+';
                    current = 0;
                } else {
                    if (operator == '+' || operator == '-') {
                        stack.push(eval.apply(operator, current, 0));
                    } else if (operator == '*' || operator == '/') {
                        stack.push(eval.apply(operator, (int) stack.pop(), current));
                    }

                    operator = ch;
                    current = 0;

                    if (ch == ')') {
                        while (!(stack.peek() instanceof Character)) {
                            current += (int) stack.pop();
                        }
                        operator = (char) stack.pop();
                    }
                }
            }

            var result = 0;
            while (!stack.isEmpty()) {
                result += (int) stack.pop();
            }
            return result;
        }
    }

    // https://leetcode.com/problems/basic-calculator-iii/description/
    static class Solution2 {
        private int evaluate(char operator, int x, int y) {
            if (operator == '+') {
                return x;
            } else if (operator == '-') {
                return -x;
            } else if (operator == '*') {
                return x * y;
            }
            return x / y;
        }

        private int solve(String s, int[] i) {
            Stack<Integer> stack = new Stack<>();
            int curr = 0;
            char previousOperator = '+';

            while (i[0] < s.length()) {
                char c = s.charAt(i[0]);

                if (c == '(') {
                    i[0]++;
                    curr = solve(s, i);
                } else if (Character.isDigit(c)) {
                    curr = curr * 10 + (c - '0');
                } else {
                    if (previousOperator == '+' || previousOperator == '-') {
                        stack.push(evaluate(previousOperator, curr, 0));
                    } else {
                        stack.push(evaluate(previousOperator, stack.pop(), curr));
                    }

                    if (c == ')') {
                        break;
                    }

                    previousOperator = c;
                    curr = 0;
                }

                i[0]++;
            }

            int result = 0;
            while (stack.size() > 0) {
                result += stack.pop();
            }
            return result;
        }

        public int calculate(String s) {
            s += "@";
            int[] i = new int[1];
            return solve(s, i);
        }
    }
}

File: src/leetcode_grind/day_341.rs
// https://leetcode.com/problems/basic-calculator-ii/description/
pub fn calculate_with_stack(s: String) -> i32 {
    let mut stack = vec![];
    let mut current = 0;
    let mut prev_operator = '+';

    fn eval(op: char, x: i32, y: i32) -> i32 {
        match op {
            '+' => return x,
            '-' => return -x,
            '*' => return x * y,
            '/' => return x / y,
            _ => panic!(),
        }
    }

    let mut s = s;
    s.push('@');

    for ch in s.chars() {
        if ch == ' ' {
            continue;
        }
        if char::is_digit(ch, 10) {
            current = current * 10 + (ch as i32 - '0' as i32);
        } else {
            if prev_operator == '*' || prev_operator == '/' {
                let x = stack.pop().unwrap();
                stack.push(eval(prev_operator, x, current));
            } else {
                stack.push(eval(prev_operator, current, 0));
            }

            prev_operator = ch;
            current = 0;
        }
    }

    stack.into_iter().sum()
}

pub fn calculate_no_stack(s: String) -> i32 {
    let n = s.len();

    let mut result = 0;
    let mut curr = 0;
    let mut prev_op = '+';
    let mut stack_top = 0; // prev num

    for (i, ch) in s.chars().enumerate() {
        if char::is_digit(ch, 10) {
            curr = curr * 10 + (ch as i32 - '0' as i32);
        }

        if !char::is_digit(ch, 10) && !char::is_whitespace(ch) || i == n - 1 {
            if prev_op == '+' || prev_op == '-' {
                result += stack_top;
                stack_top = if prev_op == '-' { -curr } else { curr };
            } else if prev_op == '*' {
                stack_top *= curr;
            } else if prev_op == '/' {
                stack_top /= curr;
            }

            prev_op = ch;
            curr = 0;
        }
    }

    result * stack_top
}